### PR TITLE
For Fenix #8791: consolidate tab counters

### PR DESCRIPTION
**Pull request by eliserichards**
[Fenix #8791](https://github.com/mozilla-mobile/fenix/issues/8791)

Todo: 
- [x] move fenix [`TabCounter`](https://github.com/mozilla-mobile/fenix/blob/master/app/src/main/java/org/mozilla/fenix/components/toolbar/TabCounter.kt) to AC 
- [x] move fenix [`TabCounterToolbarButton`](https://github.com/mozilla-mobile/fenix/blob/master/app/src/main/java/org/mozilla/fenix/components/toolbar/TabCounterToolbarButton.kt) to AC 
- [x] move fenix [`TabCounterMenu`](https://github.com/mozilla-mobile/fenix/blob/master/app/src/main/java/org/mozilla/fenix/components/toolbar/TabCounterMenu.kt) to AC (not sure about this - talking to Christian about it)
- [x] fix feature tests
- [x] make sure the feature dependencies aren't conflicting
   * `:feature-tabs`: add `:concept-menu`
   * `:ui-tabcounter`: add `:browser-menu2`, `:support-base`
- [x] replace session id checks in `TabsToolbarFeature`: https://github.com/mozilla-mobile/android-components/blob/244bd461c8b0012c14eb9280983cd9833c0df3cb/components/feature/tabs/src/main/java/mozilla/components/feature/tabs/toolbar/TabsToolbarFeature.kt#L23
- [x] replace Browser impls of `TabsToolbarFeature` with new params
- [x] make `TabCounterMenu` optional, and handle making the default version

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/ui/tabcounter/src/main/java/mozilla/components/ui/tabcounter/TabCounterMenu.kt
Comment (eliserichards) [inline components/ui/tabcounter/src/main/java/mozilla/components/ui/tabcounter/TabCounterMenu.kt:18]: Should we force the user to customize this? How can we make it both generic enough that the "plain" version will work, but that users can also customize it?

File: components/ui/tabcounter/src/main/java/mozilla/components/ui/tabcounter/TabCounterMenu.kt
Comment (eliserichards) [inline components/ui/tabcounter/src/main/java/mozilla/components/ui/tabcounter/TabCounterMenu.kt:21]: Fenix version uses the Tab mode here. Should we deal with custom vs private vs normal modes in AC?

Comment (eliserichards) [reply components/ui/tabcounter/src/main/java/mozilla/components/ui/tabcounter/TabCounterMenu.kt:21]: https://github.com/mozilla-mobile/fenix/blob/bd2c4079430673b69500f73b0519b3e3433db590/app/src/main/java/org/mozilla/fenix/components/toolbar/TabCounterMenu.kt#L30

File: components/ui/tabcounter/src/main/res/drawable/mozac_ui_tabcounter_ic_new.xml
Comment (eliserichards) [inline components/ui/tabcounter/src/main/res/drawable/mozac_ui_tabcounter_ic_new.xml:12]: + sign

File: components/ui/tabcounter/src/main/res/drawable/mozac_ui_tabcounter_ic_close.xml
Comment (eliserichards) [inline components/ui/tabcounter/src/main/res/drawable/mozac_ui_tabcounter_ic_close.xml:12]: `X` icon

File: components/ui/tabcounter/src/main/res/values/colors.xml
Comment (eliserichards) [inline components/ui/tabcounter/src/main/res/values/colors.xml:7]: Not sure if we should be passing colors and styles into the Tab Counter Menu

File: samples/browser/src/main/java/org/mozilla/samples/browser/BrowserFragment.kt
Comment (eliserichards) [inline samples/browser/src/main/java/org/mozilla/samples/browser/BrowserFragment.kt:186]: The intention would be that when the user creates their BrowserFragment they define the TabCounterMenu. Is this the right place for this?

Comment (csadilek) [reply samples/browser/src/main/java/org/mozilla/samples/browser/BrowserFragment.kt:186]: Yes, I think it's fine via the `BrowserFragment`. We're creating various other menus in A-C/Fenix in `BaseBrowserFragment` and views as well. If we're making the `TabCounterMenu` optional (see above), you actually don't have to create one here for Sample Browser, but it's good for testing :)

**Review by csadilek — COMMENTED** [review]
Wooo. Nice progress! 

I made a first pass and addressed your questions. Let me know if you anything isn't clear or you need anything else.

File: samples/browser/src/main/java/org/mozilla/samples/browser/BrowserFragment.kt
Comment (eliserichards) [reply samples/browser/src/main/java/org/mozilla/samples/browser/BrowserFragment.kt:186]: I think I'll hold off on adding it to the sample browser for now, and add it in as a followup PR 😄

File: components/ui/tabcounter/src/main/res/values/colors.xml
Comment (csadilek) [reply components/ui/tabcounter/src/main/res/values/colors.xml:7]: I think we can make this customizable later, if we need it. For now only Fenix uses it.

File: components/ui/tabcounter/src/main/java/mozilla/components/ui/tabcounter/TabCounterMenu.kt
Comment (csadilek) [inline components/ui/tabcounter/src/main/java/mozilla/components/ui/tabcounter/TabCounterMenu.kt:33]: yeah, this lambda should live in the application which will create the `TabCounterMenu` and specify/pass `onItemTapped`.

**Comment by eliserichards** [comment]
Updated todo:

- [ ] TabCounterToolbarButtonTest
- [ ] TabsToolbarFeatureTest
- [ ] TabCounterrTests

File: components/ui/tabcounter/src/main/res/layout/mozac_ui_tabcounter_layout.xml
Comment (eliserichards) [inline components/ui/tabcounter/src/main/res/layout/mozac_ui_tabcounter_layout.xml:11]: What does this `parentTag` do? Is it needed in this layout?

Comment (csadilek) [reply components/ui/tabcounter/src/main/res/layout/mozac_ui_tabcounter_layout.xml:11]: @eliserichards This is just used for the design view in Android Studio. Since this is a merge layout it can't otherwise know what this will be merged into. However, this was wrong here anyway because it should be `mozilla.components.ui.tabcounter.TabCounter` and not `org.mozilla.components.ui.tabcounter.TabCounter` which doesn't exist. In this case, I think it's fine to remove.

File: components/feature/tabs/src/main/java/mozilla/components/feature/tabs/toolbar/TabCounterToolbarButton.kt
Comment (csadilek) [inline components/feature/tabs/src/main/java/mozilla/components/feature/tabs/toolbar/TabCounterToolbarButton.kt:32]: @eliserichards we just fixed a bug in Fenix that was caused by us passing in `isPrivate` when the button gets created as opposed to looking it up dynamically. Let's remove the `isPrivate` parameter and lookup the correct tab in the flow below. Here's the change in Fenix: https://github.com/mozilla-mobile/fenix/pull/16604/commits/1e23acb8473394165581703893e79529539dc3fd

Comment (csadilek) [reply components/feature/tabs/src/main/java/mozilla/components/feature/tabs/toolbar/TabCounterToolbarButton.kt:32]: Can we bring this change in as well so we don't regress when we're switching over to this new version in A-C?

**Comment by eliserichards** [comment]
This is ready to be tested with https://github.com/mozilla-mobile/fenix/issues/8791 (branch name: `eliserichards:8791-tab-counter`)

File: components/feature/tabs/src/main/java/mozilla/components/feature/tabs/toolbar/TabCounterToolbarButton.kt
Comment (csadilek) [reply components/feature/tabs/src/main/java/mozilla/components/feature/tabs/toolbar/TabCounterToolbarButton.kt:48]: We also need to set the parameter in Sample Browser, maybe through the feature?

**Review by csadilek — CHANGES_REQUESTED** [review]
Almost there! A few more things we should work through / improve, see below.

File: components/feature/tabs/src/main/java/mozilla/components/feature/tabs/toolbar/TabCounterToolbarButton.kt
Comment (eliserichards) [reply components/feature/tabs/src/main/java/mozilla/components/feature/tabs/toolbar/TabCounterToolbarButton.kt:42]: This is only public for testing, otherwise it's private. 

I'm confused about specifying the type -- your example is the same as what I have written 🤔

Comment (csadilek) [reply components/feature/tabs/src/main/java/mozilla/components/feature/tabs/toolbar/TabCounterToolbarButton.kt:42]: > This is only public for testing, otherwise it's private.

It can't be though because you use it in your subtype in `Fenix`. If you make it private (desired) `FenixTabCounterToolbarButton` won't compile. Ideally, we shouldn't need the subtype in Fenix, see my comments there :).

> I'm confused about specifying the type 

Ah just a copy/paste error, I updated the snipped. We don't need to specify the type here.

File: components/feature/tabs/src/test/java/mozilla/components/feature/tabs/toolbar/TabCounterToolbarButtonTest.kt
Comment (eliserichards) [reply components/feature/tabs/src/test/java/mozilla/components/feature/tabs/toolbar/TabCounterToolbarButtonTest.kt:135]: I didn't think that the toolbar button was the right place to test the tab counter functionality?

Comment (eliserichards) [reply components/feature/tabs/src/test/java/mozilla/components/feature/tabs/toolbar/TabCounterToolbarButtonTest.kt:135]: But I'm happy to add them back in! 😄

Comment (csadilek) [reply components/feature/tabs/src/test/java/mozilla/components/feature/tabs/toolbar/TabCounterToolbarButtonTest.kt:135]: Yeah, let's add the tests back in. I think they're valuable.

File: components/ui/tabcounter/src/main/java/mozilla/components/ui/tabcounter/TabCounterMenu.kt
Comment (csadilek) [inline components/ui/tabcounter/src/main/java/mozilla/components/ui/tabcounter/TabCounterMenu.kt:50]: Hard-coded string :). We need to create a string for i18n, otherwise we won't have translations. We can bring in the Fenix ones.

File: components/ui/tabcounter/src/main/java/mozilla/components/ui/tabcounter/TabCounterMenu.kt
Comment (csadilek) [inline components/ui/tabcounter/src/main/java/mozilla/components/ui/tabcounter/TabCounterMenu.kt:62]: Hard-coded string :). We need to create a string for i18n, otherwise we won't have translations. We can bring in the Fenix ones.

File: components/ui/tabcounter/src/main/java/mozilla/components/ui/tabcounter/TabCounterMenu.kt
Comment (csadilek) [inline components/ui/tabcounter/src/main/java/mozilla/components/ui/tabcounter/TabCounterMenu.kt:74]: Hard-coded string :). We need to create a string for i18n, otherwise we won't have translations. We can bring in the Fenix ones.

**Review by csadilek — CHANGES_REQUESTED** [review]
Pretty much ready to land after one more round of testing. Found some hard-coded strings we need to fix before landing.

**Comment by eliserichards** [comment]
See [Fenix PR](https://github.com/mozilla-mobile/fenix/pull/16605) for test cases

**Review by csadilek — APPROVED** [review]
🚢 

We did another round of testing in sample browser, r-b and fenix. Filed two follow-ups:
https://github.com/mozilla-mobile/android-components/issues/9129
https://github.com/mozilla-mobile/android-components/issues/9130

**Pull request merged**
